README.rst
pen
===

Unofficial fork of the pen TCP/UDP load balancer (and reverse proxy) tool.

.. image:: https://travis-ci.org/myint/pen.svg?branch=master
    :target: https://travis-ci.org/myint/pen
    :alt: Build status


Installation
============

::

    $ ./bootstrap
    $ ./configure
    $ make install


Usage
=====

Usage when forwarding a single port::

    $ pen -S 1 8888 savannah.nongnu.org:80


Description
===========

This is pen, a load balancer for "simple" TCP based protocols such as
HTTP or SMTP. It allows several servers to appear as one to the
outside and automatically detects servers that are down and distributes
clients among the available servers. This gives high availability and
scalable performance.

The load balancing algorithm keeps track of clients and will try to
send them back to the server they visited the last time. The client
table has a number of slots (default 2048, settable through command-line
arguments). When the table is full, the least recently used one will
be thrown out to make room for the new one.

This is superior to a simple round-robin algorithm, which sends a client
that connects repeatedly to different servers. Doing so breaks
applications that maintain state between connections in the server,
including most modern web applications.

When pen detects that a server is unavailable, it scans for another
starting with the server after the most recently used one. That way
we get load balancing and "fair" failover for free.

Correctly configured, pen can ensure that a server farm is always
available, even when individual servers are brought down for maintenance
or reconfiguration. The final single point of failure, pen itself,
can be eliminated by running pen on several servers, using VRRP to
decide which is active.

A side-effect of the load-balancing is that several logfiles are produced,
and all accesses seem to come from the load balancer. The program
penlogd solves this problem by merging pen's log file with the ones
produced by the web servers. See penlogd(1) and penlog(1) for details.

This load balancer is known to work on FreeBSD, Linux, HP-UX and Solaris.
Other Unixes should work as well, possibly requiring trivial changes.
Success stories or problem reports are welcome.

If pen is started with the -w option, statistics are saved in html
format when a USR1 signal is received.


ChangeLog
140804	Released 0.25.0.

140701	Return UDP replies from the server to the client.

140621	UDP load balancing code restructured and bugfixed. Released 0.24.0.

140616	Released 0.23.0.

140417	In mainloop_select: When there is a pending connection, keep accepting
	up to multi_accept times *or* until EAGAIN *or* connection table is full.
	This improves performance under load.

140414	Updated GeoIP support for IPv6.

140410	Servers can have ipv6 addresses. It is possible to use a mix of ipv4
	and ipv6 servers:
	./pen -df -S 2 -r :::2222 [::1]:22 [127.0.0.1]:22

140409	In order to allow server addresses with : in them (i.e. ipv6), it is now
	possible to use square brackets around the address part of the server
	specification: [address]:port (e.g. [::1]:8080).

140408	Pen can now listen on ipv6 sockets in addition to ipv4 and unix ones.
	I.e. things like "pen ::1:2222 127.0.0.1:22" are now possible.

140402	snprintf format errors reported by Christopher Meng fixed in
	pen.c and penctl.c.
	Released 0.22.1.

140331	Updated pen manpage to clarify what the control socket does.
	Resist opening control socket running as root.
	Remove the default file name for web log.
	New feature: unix domain listening sockets.
	Released 0.22.0.

140221	Redesigned server and client structs to allow ipv6 addresses and require
	less casting (yuck) in the code.

140205	Updated penctl man page with syntax for IPv6 and GeoIP access lists.
	Fixed cosmetic signedness compiler warnings.

140204	Moved defines for ACE_IPV4 et al outside #ifdef HAVE_SSL clause.
	Otherwise pen won't compile without ssl.
	Released 0.21.1.

140204	GeoIP access lists.
	Released 0.21.0.

140120	Added "special exception" clause for linking with OpenSSL.
	Released 0.20.2.

131127	Penlog ipv6 compatible.

131126	Modernized automake configuration.
	Penctl ipv6 compatible.

131123	Updated autoconf to 2.69.
	Released 0.20.1.

131122	Updated SSL code. Protocol ssl2 removed. Default changed to tls1.
	Released 0.20.0.

131120	Added UDP patch from Zen.
	Released 0.19.0.

080517	Added patch from Debian that fixes some issues with penctl.cgi.
	Released 0.18.0.

080503	Priority based server selection algorithm.

080501	Patch from Stephen P. Schaefer fixes several issues in
	write_cfg.
	In the server_by_weight function, multiply current connections
	by WEIGHT_FACTOR to make the selection mo fine grained when the
	number of connections is small.
	Released 0.17.3.

080326	Patch from Dana Contreras: send stdio to /dev/null after
	forking.
	Fixed a bunch of cosmetic signedness compiler warnings.

070912	Added sanity check to init() to make sure that servers_max
	is large enough for the number of servers specified on the
	command line.
	Released 0.17.2.

070829	Faster string duplication courtesy of Nigel Horne.

070502	Updated automake links.

070212	Updated INSTALL with instructions for increasing the number
	of connections on Windows.

061204	Include server weight in response to "penctl servers" command.
	Set server weight on command line. Server is specified as
	host:port:maxconn:hard:weight.

060627	Bugfix by Chris Elsworth: server_by_weight would never
	consider blacklisted servers, which kept them blacklisted
	indefinitely.
	Released 0.17.1.

051230	Added code by Chris Elsworth for kqueue support. Configure
	with --with-kqueue to enable.
	Released 0.17.0.

051215	Changed the configure option for ssl to
	--with-experimental-only-ssl.
	Released 0.16.0.

040709	Added an "abuse server" for naughty clients that have been
	denied access by an acl. Works similar to the emergency server,
	use command-line option "-B host:port" to enable.

040706	Servers can be assigned different weights to account for
	differing capacity. New -W command-line option. New penctl
	commands:
	server S weight W	(assign weight to server)
	weight			(use weight for server selection)
	no weight		(do not use weight for server selection)

	Cleaned up the logic in add_client so the weighted server
	selection can be used without client tracking.

	Released 0.15.0.

040624	Some performance enhancing changes:

	New variable connections_used remembers the number of used
	slots in conns[]. It is incremented by store_conn and
	decremented by close_conn. This allows the main loop to
	only accept new connections if there are empty slots in
	conns[], which is much better than accepting the connection
	only to immediately close it because we can't handle it.

	New variable connections_last remembers the last used slot
	in conns[]. This allows us to scan for empty slots much faster
	in store_conn when there are many simultaneous connections.

	Released 0.14.0.

040622	Documented the procedure to change FD_SETSIZE on Linux
	in INSTALL.

040604	Documented the include command in the penctl manpage.

040527	Fixed SSL so it works in nonblocking mode, except that
	it doesn't work anyway.
	Moved listenfd and ctrlfd out of main.

040428	Highly experimental SSL code in pen.c. Updated manpage
	with the new options. Added https example to HOWTO.
	Released 0.13.0.

040403	Replaced signal with sigaction so connect() to nonexisting
	servers can be interrupted and not restarted.
	Released 0.12.3.

040318	Bugfix by Andreas Wrede: Running pen with -H to generate
	X-Forwarded-For headers causes POST request data sections
	to be dropped, if they are in the same packet as the POST
	header (rather than a continuation packet) and the if the
	POST data contains binary zeros.  This is almost always the
	case with Internet Explorer 6.  Mozilla and friends send
	the data of a POST request in a separate packet, so the
	problem does not show up there.

040310	Changed setlinebuf to setvbuf in penlogd.c
	Released 0.12.2.

040201	Bugfix: rewrite_request would only add the X-Forwarded-For
	header for the first request.
	Released 0.12.1.

040104	Removed the reference to mergelogs from the README and
	added one to penlogd instead.

031023	Released 0.12.0.

031022	Penlogd: sscanf would read 100 bytes + terminating nul into

test.py
#!/usr/bin/env python

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import contextlib
import os
import subprocess
import time
import unittest


PEN_PATH = os.path.join(os.path.dirname(__file__), 'pen')


class Test(unittest.TestCase):

    def test_reverse_proxy(self):
        with pen(['-f', '-S', '1', '8888', 'savannah.nongnu.org:80']):
            curl = subprocess.Popen(['curl',
                                     '--silent',
                                     'http://127.0.0.1:8888'],
                                    stdout=subprocess.PIPE)
            self.assertIn('GNU', curl.communicate()[0].decode('utf-8'))
            self.assertEqual(0, curl.returncode)

    def test_handle_unwritable_file_gracefully(self):
        self.assertEqual(
            1,
            subprocess.call([PEN_PATH, '-f',
                             '-F', '/',
                             '-o', 'write', '/',
                             '8888', 'savannah.nongnu.org:80'],
                            stderr=subprocess.PIPE))


@contextlib.contextmanager
def pen(arguments):
    process = subprocess.Popen([PEN_PATH] + arguments)

    time.sleep(2)

    try:
        yield process
    finally:
        process.kill()


if __name__ == '__main__':
    unittest.main()


test.bash
#!/bin/bash

set -e
set -x

output_filename='test.html'
pid_filename='pen.pid'

./pen -p "$pid_filename" -S 1 8888 savannah.nongnu.org:80

while [ ! -e "$pid_filename" ]
do
    sleep .1
done

clean()
{
    rm -f "$output_filename"

    pen_pid=$(cat "$pid_filename")
    kill "$pen_pid"
}
trap clean EXIT

curl -o "$output_filename" http://127.0.0.1:8888
grep -q GNU "$output_filename"

echo -e '\x1b[01;32mOkay\x1b[0m'


penctl.c
/*
   Copyright (C) 2000-2014  Ulric Eriksson <ulric@siag.nu>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.
*/

#include "config.h"

#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <netdb.h>
#include <sys/types.h>
#include <netinet/in.h>
//#include <arpa/inet.h>
#include <signal.h>
#include <stdlib.h>
#include <stdarg.h>
#include <ctype.h>
#include <time.h>
#ifdef TIME_WITH_SYS_TIME
#include <sys/time.h>
#endif
#include <sys/wait.h>
#include <sys/resource.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <syslog.h>
#include <string.h>

#include "config.h"

static void error(char *fmt, ...)
{
	char b[4096];
	va_list ap;
	va_start(ap, fmt);
	vsnprintf(b, sizeof b, fmt, ap);
	fprintf(stderr, "%s\n", b);
	va_end(ap);
	exit(EXIT_FAILURE);
}

static void alarm_handler(int dummy)
{
	(void)dummy;
}

static void usage(void)
{
	printf("usage: penctl host:port command\n");
	exit(0);
}

static int open_unix_socket(char *path)
{
	int n, fd;
	struct sockaddr_un serv_addr;

	fd = socket(PF_UNIX, SOCK_STREAM, 0);
	if (fd < 0) error("error opening socket");
	memset(&serv_addr, 0, sizeof serv_addr);
	serv_addr.sun_family = AF_UNIX;
	snprintf(serv_addr.sun_path, sizeof serv_addr.sun_path, "%s", path);
	n = connect(fd, (struct sockaddr *)&serv_addr, sizeof serv_addr);
	if (n == -1) {
		error("error connecting to server");
	}
	return fd;
}

static int open_socket(char *addr, char *port)
{
	int fd = -1;
#if 0
	struct sockaddr_in serv_addr;
	struct hostent *h;
	struct in_addr a;
	struct servent *s;
#else
	struct addrinfo *ai;
	struct addrinfo hints;
	struct addrinfo *runp;
#endif
	int n;
#if 0
	h = gethostbyname(addr);
	if (h == NULL) error("unknown or invalid address %s\n", addr);
	memcpy(&a, h->h_addr, h->h_length);
	s = getservbyname(port, "tcp");
	if (s == NULL) po = atoi(port);
	else po = ntohs(s->s_port);
	fd = socket(AF_INET, SOCK_STREAM, 0);
#else
	memset(&hints, 0, sizeof(hints));
	hints.ai_flags = AI_ADDRCONFIG;
	hints.ai_socktype = SOCK_STREAM;
	n = getaddrinfo(addr, port, &hints, &ai);
	if (n != 0) {
		error("getaddrinfo: %s", gai_strerror(n));
	}
	runp = ai;
	/* only using first result; should try all */
	fd = socket(runp->ai_family, runp->ai_socktype, runp->ai_protocol);
#endif

	if (fd < 0) error("error opening socket");
#if 0
	serv_addr.sin_family = AF_INET;
	serv_addr.sin_addr.s_addr = a.s_addr;
	serv_addr.sin_port = htons(po);
#endif
	signal(SIGALRM, alarm_handler);
#if 0
	n = connect(fd, (struct sockaddr *)&serv_addr, sizeof serv_addr);
#else
	n = connect(fd, runp->ai_addr, runp->ai_addrlen);
#endif
	alarm(0);
	if (n == -1) {
		error("error connecting to server");
	}
	return fd;
}

int main(int argc, char **argv)
{
	int i, fd, n;
	char b[1024], *p;

	if (argc < 3) {
		usage();
	}

	if (strchr(argv[1], '/')) {
		fd = open_unix_socket(argv[1]);
	} else {
		n = 1+strlen(argv[1]);	/* one for \0 */
		if (n > sizeof b) error("Overlong arg '%s'", argv[1]);
		strcpy(b, argv[1]);
		/* We need the *last* : to allow such arguments as ::1:10080
		   if pen's control port is ipv6 localhost:10080 */
		p = strrchr(b, ':');
		if (p == NULL) error("no port given");

		*p++ = '\0';

		fd = open_socket(b, p);
	}

	n = 2+strlen(argv[2]);	/* one for \n, one for \0 */
	if (n > sizeof b) error("Overlong arg '%s'", argv[2]);
	strcpy(b, argv[2]);
	for (i = 3; argv[i]; i++) {
		n = n+1+strlen(argv[i]);
		if (n > sizeof b) error("Overlong arg '%s'", argv[i]);
		strcat(b, " ");
		strcat(b, argv[i]);
	}
	strcat(b, "\n");
	n = write(fd, b, strlen(b));
	if (n == -1) error("error writing to socket");
	for (;;) {
		n = read(fd, b, sizeof b);
		if (n == 0) break;
		if (n == -1) error("error reading from socket");
		n = write(1, b, n);
		if (n == -1) error("error writing to stdout");
	}
	close(fd);

	return 0;
}


penlogd.c
/*
   Copyright (C) 2002-2014  Ulric Eriksson <ulric@siag.nu>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <time.h>
#include <signal.h>
#include <stdarg.h>
#include <syslog.h>
#include <pwd.h>
#include <errno.h>
#include <ctype.h>

#define PEN_MAX 1000	/* make that at least 1000 for prod */

/* This structure is statically allocated, which is wasteful but saves us
   from frequest memory allocation.
*/
static struct penlog {
	struct in_addr addr;
	char client[100];
	char request[100];
} *penlog;

static int pen_n = 0;		/* next slot in buffer */

static int debuglevel = 0;
static int foreground = 0;
static int unbuffer = 0;
static char *logfile = NULL;
static FILE *logfp;
static char *pidfile = NULL;
static int loopflag = 1;
static int do_restart_log = 0;
static int pen_max = PEN_MAX;
static char *user = NULL, *jail = NULL;

static struct sigaction hupaction, termaction;

static void debug(char *fmt, ...)
{
	time_t now;
	struct tm *nowtm;
	char nowstr[80];
	char b[4096];
	va_list ap;
	va_start(ap, fmt);
	vsnprintf(b, sizeof b, fmt, ap);
	now=time(NULL);
	nowtm = localtime(&now);
	strftime(nowstr, sizeof(nowstr), "%Y-%m-%d %H:%M:%S", nowtm);
	if (foreground) {
		fprintf(stderr, "%s: %s\n", nowstr, b);
	} else {
		openlog("penlogd", LOG_CONS, LOG_USER);
		syslog(LOG_DEBUG, "%s\n", b);
		closelog();
	}
	va_end(ap);
}

static void error(char *fmt, ...)
{
	char b[4096];
	va_list ap;
	va_start(ap, fmt);
	vsnprintf(b, sizeof b, fmt, ap);
	if (foreground) {
		fprintf(stderr, "%s\n", b);
	} else {
		openlog("penlogd", LOG_CONS, LOG_USER);
		syslog(LOG_ERR, "%s\n", b);
		closelog();
	}
	va_end(ap);
	exit(1);
}

static void restart_log(int dummy)
{
	(void)dummy;

	do_restart_log = 1;
	sigaction(SIGHUP, &hupaction, NULL);
}

static void quit(int dummy)
{
	(void)dummy;

	loopflag = 0;
}

static void store_web(char *b, int n, struct in_addr addr)
{
	char request[1024];
	char *p, *q;
	int i, m;

	b[n] = '\0';
	if (debuglevel > 1) debug("store_web(%s, %d)", b, n);
	p = strchr(b, '"');
	if (p == NULL) {
		debug("bogus web line %s", b);
		return;
	}
	p++;
	q = strchr(p, '"');
	if (q == NULL) {
		debug("bogus line %s", b);
		return;
	}
	memcpy(request, p, q-p);
	request[q-p] = '\0';
	if (q-p < 100) m = q-p;
	else m = 100;
	i = pen_n-1;
	if (i < 0) i = pen_max-1;
	while (i != pen_n) {
		if (penlog[i].request[0] &&
		    addr.s_addr == penlog[i].addr.s_addr &&
		    !strncmp(request, penlog[i].request, m)) {
			break;
		}
		i--;
		if (i < 0) i = pen_max-1;
	}

	if (i == pen_n) {	/* no match */
		fwrite(b, 1, n, logfp);
	} else {
		fputs(penlog[i].client, logfp);
		p = strchr(b, ' ');
		if (p == NULL) {
			if (debuglevel) debug("Ugly");
			return;
		}
		fwrite(p, 1, n-(p-b), logfp);
	}
}

static void store_pen(char *b, int n)
{
	char client[100], server[100], request[100];
#ifdef HAVE_INET_ATON
	struct in_addr addr;
#else
	struct hostent *hp;
#endif

	b[n] = '\0';
	if (sscanf(b, "+ %99[^ ] %99[^ ] %99[^\n]",
		   client, server, request) != 3) {
		debug("discarding bogus pen line %s", b);
		return;
	}
	if (debuglevel > 1)
		debug("store_pen(%i: %s, %s, %s)",pen_n, client, server, request);

#ifdef HAVE_INET_ATON
	if (inet_aton(server, &addr) == 0) {
		debug("bogus address %s", server);
		return;
	}
	penlog[pen_n].addr = addr;
#else
	hp = gethostbyname(server);
	memcpy(&penlog[pen_n].addr, hp->h_addr, hp->h_length);
#endif


	strncpy(penlog[pen_n].client, client, 100);
	strncpy(penlog[pen_n].request, request, 100);
	pen_n++;
	if (pen_n >= pen_max) pen_n = 0;
}

static void usage(void)
{
	printf("Usage:\n"
	       "  penlogd [options] port\n"
	       "\n"
	       "  -d        debugging on\n"
	       "  -f        stay in foreground\n"
	       "  -j dir    run in chroot\n"
	       "  -l file   write log to file\n"
	       "  -b        unbuffer output (Testing Only!)\n"
	       "  -n N      number of pen log entries to cache [1000]\n"
	       "  -p file   write pid to file\n"
	       "  -u user   run as alternative user\n");
	exit(0);
}

static void background(void)
{
#ifdef HAVE_DAEMON
	daemon(0, 0);
#else
	int childpid;
	if ((childpid = fork()) < 0) {
		error("Can't fork");
	} else {
		if (childpid > 0) exit(0);	/* parent */
	}
	setsid();
	signal(SIGCHLD, SIG_IGN);
#endif
}

static int options(int argc, char **argv)
{
	int c;

	while ((c = getopt(argc, argv, "j:l:n:p:u:dfb")) != -1) {
		switch (c) {
		case 'd':
			debuglevel++;
			break;
		case 'b':
			unbuffer = 1;
			break;
		case 'f':
			foreground = 1;
			break;
		case 'j':
			jail = optarg;
			break;
		case 'l':
			logfile = optarg;
			break;
		case 'n':
			pen_max = atoi(optarg);
			break;
		case 'p':
			pidfile = optarg;
			break;
		case 'u':
			user = optarg;
			break;
		default:
			usage();
		}
	}

	return optind;
}

int main(int argc, char **argv)
{
	struct passwd *pwd = NULL;
	struct sockaddr_in a;
	socklen_t len;
	int ld, p;
	char b[1024];

	int n = options(argc, argv);
	argc -= n;
	argv += n;

	if (argc < 1 || (p = atoi(argv[0])) == 0) {
		usage();
		exit(0);
	}

	penlog = malloc(pen_max * sizeof *penlog);
	if (!penlog) error("Can't allocate penlog");

	if (!foreground) background();

	if (user) {
		if (debuglevel) debug("Run as user %s", user);
		pwd = getpwnam(user);
		if (pwd == NULL) error("Can't getpwnam(%s)", user);
	}
	if (jail) {
		if (debuglevel) debug("Run in %s", jail);
		if (chroot(jail) == -1) error("Can't chroot(%s)", jail);
	}
	if (pwd) {
		if (setuid(pwd->pw_uid) == -1)
			error("Can't setuid(%d)", (int)pwd->pw_uid);
	}
	if (logfile) {
		if (debuglevel) debug("Logging to %s", logfile);
		logfp = fopen(logfile, "a");
		if (!logfp) error("Can't open logfile %s", logfile);
		if (unbuffer) setvbuf(logfp, (char *)NULL, _IOLBF, 0);
	} else {
		if (debuglevel) debug("Logging to stdout");
		logfp = stdout;
	}
	if (pidfile) {
		FILE *pidfp = fopen(pidfile, "w");
		if (debuglevel) {
			debug("Writing pid %d to %s",
				(int)getpid(), pidfile);
		}
		if (!pidfp) error("Can't create pidfile %s", pidfile);
		fprintf(pidfp, "%d", (int)getpid());
		fclose(pidfp);
	}

	if ((ld = socket(PF_INET, SOCK_DGRAM, 0)) < 0) {
		error("Problem creating socket");
	}

	a.sin_family = AF_INET;
	a.sin_addr.s_addr = htonl(INADDR_ANY);
	a.sin_port = htons(p);

	if (bind(ld, (struct sockaddr *) &a, sizeof(a)) < 0) {
		error("Problem binding");
	}

	len = sizeof a;
	if (getsockname(ld, (struct sockaddr *) &a, &len) < 0) {
		error("Error getsockname");
	}

	hupaction.sa_handler = restart_log;
	sigemptyset(&hupaction.sa_mask);
	hupaction.sa_flags = 0;
	sigaction(SIGHUP, &hupaction, NULL);
	termaction.sa_handler = quit;
	sigemptyset(&termaction.sa_mask);
	termaction.sa_flags = 0;
	sigaction(SIGTERM, &termaction, NULL);

	loopflag = 1;

	if (debuglevel) debug("Enter main loop\n");

	while (loopflag) {
		if (do_restart_log) {
			if (debuglevel) debug("Reopen log file %s", logfile);
			if (logfp != stdout) {
				fclose(logfp);
				logfp = fopen(logfile, "a");
				if (!logfp) error("Can't open %s", logfile);
				if (unbuffer) setvbuf(logfp, (char *)NULL, _IOLBF, 0);
			}
			do_restart_log = 0;
		}
		n = recvfrom(ld, b, sizeof b, 0, (struct sockaddr *) &a, &len);

		if (n < 0) {
			if (errno != EINTR)
				perror("Error receiving data");
			continue;
		}
		b[n] = 0;
#if 0	/* what on earth would send us something like that? */
		if (strstr(b, "Host:")) {
			continue;	/* do nothing */
#endif
		if (b[0] == '+') {
			store_pen(b, n);
		} else {
			store_web(b, n, a.sin_addr);
		}
	}

	if (debuglevel) debug("Exit main loop");

	if (logfp) fclose(logfp);
	if (pidfile) unlink(pidfile);
	return 0;
}
